### ISSUE TITLE
Rewrite Builder shell scripts in Go

### ISSUE THREAD

**Issue by technosophos** [deis/deis#3836]
The current builder is written in shell scripts. While some of the shell should remain, the majority of it should be reimplemented in Go for the following reasons:

- We get finer grained control over interoperability with etcd and other services
- We can write good tests!
- The platform becomes more consistent.

This is part of the overarching milestone described in #3758.

**Comment by technosophos** [comment]
_I'm going to paste in here a slightly revised snippet of a comment I made in a great conversation with @kalbasit. I think it's important to bubble this up._

_Background: My current prototype uses a chain-of-command pattern with dependency injection to create a composable architecture. Or, in plain English, it maps a named task ("boot") to a series of individual (reusable) steps. What I'm pasting below explains my rationale._

[L]et me hit on the big design goals of this rewrite:
1. Replace a bunch of shell with Go
2. Make everything really, really testable
3. Hopefully work toward a component-based system that we can re-use

The first is easy to accomplish.

For the other two, here's what I am thinking: Basically, builder's boot process runs a sequence of individual steps. Some of these steps are really specific, but most are fairly generic. Read a bunch of environment variables. Send a request to etcd. Start a system service.

So what I'm doing is taking a system that provides the ability to run a sequence of commands, and writing a bunch of re-usable commands. Since every command follows the same API contract, the system is very composable. It's designed for dependency injection, which makes writing good tests very easy as well -- whether unit tests or functional tests. And at the top level, it should be really easy for a new developer to look at the route description and figure out exactly what the process is.

Defining a route (task list, chain) looks like this:
```go
	reg.Route("boot", "Boot the builder.").
		// Get env vars
		Does(env.Get, "vars").
		Using("HOST").WithDefault("127.0.0.1").
		Using("ETCD_PORT").WithDefault("4001").
		Using("ETCD").WithDefault("http://$HOST:$ETCD_PORT").
		Using("ETCD_PATH").WithDefault("/deis/builder").
		Using("ETCD_TTL").WithDefault("20").
		// Are there cases where we need to support more than one URL?
		Does(etcd.Client, "client").
		Using("url").WithDefault("http://localhost:4001").From("cxt:ETCD").
		// See if etcd is awake. This is the legacy way of doing it.
		Does(etcd.Ls, "ls").
		Using("client").From("cxt:client").
		Using("retries").WithDefault(20).
		// Sleep while etcd clears its stale values. Do we really need to do this?
		Does(Sleep, "_").Using("duration").WithDefault(21*time.Second).
		// Generate the name of the users directory.
		Does(fmt.Sprintf, "newdir").
		Using("format").WithDefault("%s/users").
		Using("0").From("cxt:ETCD_PATH").
		// Create the users directory in etcd.
		Does(etcd.MakeDir, "mkdir").
		Using("path").From("cxt:newdir").
		Using("client").From("cxt:client")
```

Route `boot` does the following:
- Gets env vars and puts them in the context.
- Creates an etcd client 
- Runs an etcd `ls` command
- Sleeps for 21 seconds
- Creates a new directory name
- Makes a new directory

(This is the sequence that the current builder shell script runs)

Provided that every command is properly documented, it should be easy to then dive into the specifics of the code.

There are three things I am not particularly thrilled with:

1. `cookoo.Params` objects require a little extra documenting (or code-reading).
2. There are a lot of type assertions required in this approach.
3. `go fmt` is really strict on indentation, so more decorative whitespacing is not allowed, e.g.:
```
reg.Route(foo).
    Does(bar).
        Using(baz).WithDefault("OH HAI")
```

But these feel like minor annoyances when compared with the benefit of building a composed system like this.

_The library I'm using in the example above is `github.com/Masterminds/cookoo`._

**Comment by kalbasit** [comment]
_Pasting also my comment here, let's keep talking here I guess_

> `cookoo.Params` objects require a little extra documenting (or code-reading).

Exactly what I meant, it took me a little while to understand what the above code does. It's not very idiomatic. However, it might be beneficial to do it this way, I guess only time will tell. I'll keep an eye out for it.

**Comment by technosophos** [comment]
The new format for the routes should look much more familiar:

```go
	reg.AddRoute(cookoo.Route{
		Name: "boot",
		Help: "Boot the builder",
		Does: cookoo.Tasks{

			// ENV: Make sure the environment is correct.
			cookoo.Cmd{
				Name: "vars",
				Fn:   env.Get,
				Using: []cookoo.Param{
					{Name: "HOST", DefaultValue: "127.0.0.1"},
					{Name: "ETCD_PORT", DefaultValue: "4001"},
					{Name: "ETCD_PATH", DefaultValue: "/deis/builder"},
					{Name: "ETCD_TTL", DefaultValue: "20"},
				},
			},
			// ETCD: Make sure Etcd is running, and do the initial population.
			cookoo.Cmd{
				Name:  "client",
				Fn:    etcd.CreateClient,
				Using: []cookoo.Param{{Name: "url", DefaultValue: "http://127.0.0.1:4001", From: "cxt:ETCD"}},
			},
			cookoo.Cmd{
				Name: "etcdup",
				Fn:   etcd.IsRunning,
				Using: []cookoo.Param{
					{Name: "client", From: "cxt:client"},
					{Name: "count", DefaultValue: 20},
				},
			},
			cookoo.Cmd{
				Name:  "-",
				Fn:    Sleep,
				Using: []cookoo.Param{{Name: "duration", DefaultValue: 21 * time.Second}},
			},
// ...
```

**Comment by mboersma** [comment]
Closed by #4273.
